### PR TITLE
test/helpers: gather logs from all pods

### PR DESCRIPTION
**Pull request by ianvernon**
Add a new function, GeneratePodLogGatheringCommands, which generates the
commands to gather logs, and the filenames which said logs will be stored in,
for all pods running in Kubernetes. This function will run if a test in the
Kubernetes test suite fails. If a container in a pod has a nonzero restart
count, the function will also generate the commands to get the logs for the
previous instance of the container.

Signed-off by: Ian Vernon <ian@cilium.io>

Fixes: #4463 

Example of the resultant logs from the test `K8sValidatedKafkaPolicyTest`. I induced a failure by making a false assertion ( `Expect(false).To(BeTrue())` ), and then listed all of the logs files in the `dafault` namespace along with the first 10 lines of each:

```
$ fls=$(ls default-*.log) && for fl in $fls ; do echo "**********first 10 lines of $fl**********"; head -n 10 $fl ; echo ; done
**********first 10 lines of default-empire-backup-74547b5f6f-7cx29-empire-backup.log**********

**********first 10 lines of default-empire-hq-6c49cd55fc-vlcns-empire-hq.log**********

**********first 10 lines of default-empire-outpost-8888-5d8d98fb99-mp7zv-empire-outpost-8888.log**********

**********first 10 lines of default-empire-outpost-9999-6666dfb499-zrdhk-empire-outpost-9999.log**********

**********first 10 lines of default-kafka-broker-79cd7b588c-hs5b7-kafka.log**********
2018-06-12T05:34:58.478186559Z waiting for kafka to be ready
2018-06-12T05:34:58.505583426Z Excluding KAFKA_HOME from broker config
2018-06-12T05:34:58.513350606Z [Configuring] 'advertised.host.name' in '/opt/kafka/config/server.properties'
2018-06-12T05:34:58.518953831Z [Configuring] 'zookeeper.connection.timeout.ms' in '/opt/kafka/config/server.properties'
2018-06-12T05:34:58.53447268Z [Configuring] 'port' in '/opt/kafka/config/server.properties'
2018-06-12T05:34:58.553196761Z [Configuring] 'broker.id' in '/opt/kafka/config/server.properties'
2018-06-12T05:34:58.582277426Z [Configuring] 'zookeeper.session.timeout.ms' in '/opt/kafka/config/server.properties'
2018-06-12T05:34:58.584142463Z Excluding KAFKA_VERSION from broker config
2018-06-12T05:34:58.598121677Z [Configuring] 'zookeeper.connect' in '/opt/kafka/config/server.properties'
2018-06-12T05:34:58.614325299Z [Configuring] 'log.dirs' in '/opt/kafka/config/server.properties'

**********first 10 lines of default-zookeeper-767d644449-gcvzc-zookeeper.log**********
2018-06-12T05:34:36.193523181Z ZOOKEEPER_ID=1
2018-06-12T05:34:36.193779814Z tickTime=2000
2018-06-12T05:34:36.19398794Z initLimit=10
2018-06-12T05:34:36.194162056Z syncLimit=5
2018-06-12T05:34:36.194332216Z maxClientCnxns=60
2018-06-12T05:34:36.194473803Z autopurge.snapRetainCount=3
2018-06-12T05:34:36.194630197Z autopurge.purgeInterval=0
2018-06-12T05:34:36.208606949Z ZooKeeper JMX enabled by default
2018-06-12T05:34:36.21088226Z Using config: /opt/zookeeper/bin/../conf/zoo.cfg
2018-06-12T05:34:37.36295274Z 2018-06-12 05:34:37,350 [myid:] - INFO  [main:QuorumPeerConfig@134] - Reading configuration from: /opt/zookeeper/bin/../conf/zoo.cfg
```

Marked as high priority view so we can better debug application-related failures in the CI.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4476)

### CODE REVIEW

**Comment by ianvernon** [comment]
test-me-please

File: test/helpers/kubectl.go [hunk 1]
```diff
@@ -1197,6 +1205,33 @@ func (kub *Kubectl) GatherLogs() {
...
+	}
+	pods, err := kub.GetAllPods()
+	if err != nil {
+		kub.logger.WithError(err).Error("Unable to get pods from Kubernetes via kubectl")
+	}
+
+	for _, pod := range pods {
+		for _, containerStatus := range pod.Status.ContainerStatuses {
+			logCmd := fmt.Sprintf("%s -n %s logs --timestamps %s -c %s", KubectlCmd, pod.Namespace, pod.Name, containerStatus.Name)
+			logfileName := fmt.Sprintf("%s-%s-%s.log", pod.Namespace, pod.Name, containerStatus.Name)
```
Comment (eloycoto) [inline test/helpers/kubectl.go]: Minor, could the name be something like: `pod-%s-%s-%s.log` so all the pods logs are in the same order ;-)

**Review by eloycoto — DISMISSED** [review]

File: test/helpers/kubectl.go (hunk 1 above)
Comment (ianvernon) [reply test/helpers/kubectl.go]: Done.

**Comment by ianvernon** [comment]
test-me-please

**Review by manalibhutiyani — APPROVED** [review]

**Comment by ianvernon** [comment]
Build has passed, and I have addressed @eloycoto 's one concern, which was trivial to address. As such, to unblock CI debugging, I am merging.

**Pull request merged**